[crypto] Extend the OperationalKeystore API to allow migration

ArekBalysNordic opened the pull request:
Implemented a mechanism to detect all existing mbedTLS-related operational keys stored in non-secure storage and move them to the PSA ITS secure storage.

After the migration all keys will be removed from non-secure storage.

Damian-Nordic commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -209,5 +252,56 @@ void PSAOperationalKeystore::ReleasePendingKeypair()
+    Crypto::SensitiveDataBuffer<TLV::EstimateStructOverhead(sizeof(uint16_t), P256SerializedKeypair::Capacity())> buf;
+    uint16_t keySize = static_cast<uint16_t>(buf.Capacity());
+
+    for (FabricIndex index = 0; index < CHIP_CONFIG_MAX_FABRICS; index++)
```
This is wrong. Fabric index can be any value between kMinValidFabricIndex and kMaxValidFabricIndex, so it can be bigger than CHIP_CONFIG_MAX_FABRICS. I think one option is not to call this function from `Init`, but from Server after initializing FabricTable (and then get the list of existing fabrics from FabricTable).

bzbarsky-apple commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -17,13 +17,22 @@
 namespace Crypto {
 
+namespace {
+// Tags for our operational keypair storage copied from PersistentStorageOperationalKeystore.cpp
```
Shouldn't there be comments in PersistentStorageOperationalKeystore about how if something there changes this code needs to be updated?

bzbarsky-apple commented on src/crypto/PSAOperationalKeystore.h:
```diff
@@ -19,13 +19,29 @@
+    /**
+     * @brief Initialize the PSA Operational Keystore and move all sensitive keys from storage to ITS secure storage.
+     *
+     * After moving all sensitive keys to ITS, the old one stored in the storage will be removed and not available anymore.
```
This needs to more clearly document that this can fail partway through, after having moved some of the keys over, right?  And what the exact behavior is in that case.

ArekBalysNordic replied:
I've changed the way to migrate those keys. Now the keys are migrated on the HasOpKeypairForFabric method call, so in fact on CASE session establishment. Thanks to that when one of the migrations finished with an error, the other ones for other fabrics will be working and it won't break the device, like previously.

ArekBalysNordic commented on src/app/server/Server.cpp:
```diff
@@ -655,7 +655,11 @@ void Server::ResumeSubscriptions()
 
 KvsPersistentStorageDelegate CommonCaseDeviceServerInitParams::sKvsPersistenStorageDelegate;
+#if CHIP_CRYPTO_PSA
+PSAOperationalKeystore CommonCaseDeviceServerInitParams::sPSAOperationalKeystore;
```
Yes, that's right, but there are many `CommonCaseDeviceServerInitParams` initialized this way, and the `sPersistentStorageOperationalKeystore` has been used there already. I've just added another option to initialize `sPSAOperationalKeystore` instead of `sPersistentStorageOperationalKeystore` as `operationalKeystore` if the PSA backend is selected.

bzbarsky-apple replied:
Again, my question is: Are there many in-tree examples that use PSA here?

Damian-Nordic replied:
Boris, I switched nRF Connect unit tests to PSA a few days ago. Also, OpenIOT runs tests with PSA. Apart from that, no in-tree examples use PSA as of now, but all nRF examples will switch to PSA within ~2 months.

bzbarsky-apple replied:
OK, so it's not like there are "many examples that would be hard to update to not use CommonCaseDeviceServerInitParams" right now?  And if nRF switches all examples, we should think about how to do that without proliferating the deprecated CommonCaseDeviceServerInitParams even more....

Damian-Nordic replied:
It's not hard to not to update `CommonCaseDeviceServerInitParams` right now but I'm not sure in which direction we're planning to steer away from that before we switch nRF examples to PSA. If we remove `CommonCaseDeviceServerInitParams`, does it mean that every time we add a new thing to `ServerInitParams` we have to add the same boilerplate to each in-tree example? And have all device manufacturers read about interdependencies between the components, such as "when you use PSA crypto, you must not use an operational key store that writes keys to KVS" etc. The point of changing the default for PSA was that the existing default would not work at all with PSA due to a different key format.

Note that for things like crypto, most device manufacturers have no reason to provide their own crypto implementations. They just want to use whatever platform vendors like us recommend. I can imagine no better way of making everyone's life easier than to provide default implementations that were tested to work with each other. Everyone is free to tinker with the params once they learn Matter SDK and they need to.

tcarmelveilleux commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -209,5 +254,53 @@ void PSAOperationalKeystore::ReleasePendingKeypair()
     mIsPendingKeypairActive = false;
 }
 
+CHIP_ERROR PSAOperationalKeystore::MoveOpKeysFromKVSToITS(FabricIndex fabricIndex) const
```
This should not mention "KVS", but rather "PersistentStorage" (which may not be backed by Kvs directly).

tcarmelveilleux commented on src/app/server/Server.h:
```diff
@@ -214,10 +218,15 @@ struct CommonCaseDeviceServerInitParams : public ServerInitParams
         if (this->operationalKeystore == nullptr)
         {
+#if CHIP_CRYPTO_PSA
+            ReturnErrorOnFailure(sPSAOperationalKeystore.Init(this->persistentStorageDelegate));
```
This will fail to work for any client using a custom ServerInitParams. I guess that's OK, but bear in mind some existing products will then miss the migration.

Also, failing on migration means that the product will error-out Server::Init, which *may lead to bootlooping*. It is extremely risky to do "database migration" in a way that errors are not passively logged/ignored. It means that if anything arises which was not seen in the field, products that upgrade to this new code can get bricked. I would recommend leaving it to individual products to opt-in to this migration, or alternatively, make sure that it cannot cause a failure of Server Init.

tcarmelveilleux commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -209,5 +254,53 @@ void PSAOperationalKeystore::ReleasePendingKeypair()
+
+            VerifyOrReturnError(keyData.size() == (kP256_PrivateKey_Length + kP256_PublicKey_Length), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+            // Migrate the operational keys
```
All code in this method, above, should probably be using public APIs of PersistentStorageOperationalKeystore, rather than partially re-implement  all loading here, and then do the integration work below as well.

In other words:

- Don't reimplement the reading of the keys from persistent storage, rely on PersistentStorageOperationalKeystore.
- You can keep the code in this block in the method, as it is PSA-specific

tcarmelveilleux commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -209,5 +254,53 @@ void PSAOperationalKeystore::ReleasePendingKeypair()
+        }
+        else
+        {
+            ReturnErrorOnFailure(mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::FabricOpKey(fabricIndex).KeyName()));
```
Why fail on deletion failure? Although it shouldn't happen, isn't it safer to just tolerate it, now that the transition is done?

tcarmelveilleux commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -209,5 +254,53 @@ void PSAOperationalKeystore::ReleasePendingKeypair()
+    Crypto::SensitiveDataBuffer<TLV::EstimateStructOverhead(sizeof(uint16_t), P256SerializedKeypair::Capacity())> buf;
+    uint16_t keySize = static_cast<uint16_t>(buf.Capacity());
+
+    if (CHIP_NO_ERROR ==
```
This is complex side-effecting logic, with multiple error paths. However, I see no unit tests at all. Please add unit tests that run on POSIX with a POSIX version of PSA API.

tcarmelveilleux commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -102,6 +113,9 @@ bool PSAOperationalKeystore::HasOpKeypairForFabric(FabricIndex fabricIndex) cons
     VerifyOrReturnError(IsValidFabricIndex(fabricIndex), false);
 
+    // check if the key exists in the PersistentStorageOperationalKeystore exist, and if so move it to the PSA ITS secure storage.
+    VerifyOrReturnError(CHIP_NO_ERROR == MoveOpKeysFromKVSToITS(fabricIndex), false);
```
On failure, this can leave the device unable to be reached on a given fabric.

All of the translation should be done *before* starting to process anything, not everytime HasOpKeypairForFabric is called.

I would recommend making a separate building block that a public method "MoveOpKeysFromPersistentStorageToITS" or similar name, that can be called *on purpose, by choice* by a given product, to proceed with the migration of all fabrics at once, when the FabricTable has been initialized, such as after Server Init but before processing any incoming connections. Perhaps there is a race with subscription resumption, and it has to be done with product knowledge of storage in a way that can be done outside Server::Init(), but the idea is the same: having this behavior by default for all adds a large amount of untested risk to all integrators who are making use of PSA, but had previously commissioned devices (e.g. inherit this new path via a firmware update).

tcarmelveilleux reviewed (changes requested):
- [ ] Add unit test coverage
- [ ] Find a way that this database migration is not forced by default on all products. The migration should be a single call that is done *on purpose* by a product willing to take the risk, at a point where they can try to recover on failure (e.g. factory-reset on failure, and/or error-out to a user-facing behavior like displayed error, LED blink code, etc).

ArekBalysNordic commented:
> * [ ]  Add unit test coverage
> * [ ]  Find a way that this database migration is not forced by default on all products. The migration should be a single call that is done _on purpose_ by a product willing to take the risk, at a point where they can try to recover on failure (e.g. factory-reset on failure, and/or error-out to a user-facing behavior like displayed error, LED blink code, etc).

Thanks for the review. I will write unit tests and propose a new solution.

ArekBalysNordic commented:
I've changed my way of migrating. Now, the process takes place in the Server init method, and if the key related to a given fabric is not migrated then the new event is emitted, so the platform vendors can react on that as they wish, for example performing a factory reset. This solution never causes a boot loop in Server init, and the migration is still optional. Moreover, I extended the `OperationalKeystore API` to provide an Export and Migration method, however migration from PSA ITS to KVS is not possible because exporting is forbidden for the persistent operational keys stored in ITS. I left the implementation of the Export method in the PSA Operational Keystore when `CHIP_CONFIG_TEST` is selected because we need to verify whether the psa_export_key returns an error. From the device's perspective, there can be just return `CHIP_ERROR_NOT_SUPPORTED`. According to unit tests, I've added tests to both `PersistentStorage` and `PSAPersistentStorage` to cover migration and exporting keys.

@tcarmelveilleux @bzbarsky-apple What do you think about these changes? 

I know that any of the platforms do not use PSA currently, but we work intensively to move from mbedTLS to PSA crypto backend in the nrfconnect platform, and I guess that change will be helpful for many vendors in the future, not only for migration between mbedTLS and PSA but also other future `OperationalKeystore` implementations. As it is not used by any platform yet, I've covered most cases in the unit tests.

ArekBalysNordic commented:
As we don't have Nrfconnect examples ported to PSA yet, we will re-think how to use the migration in the platform instead of calling it in the Server Init. For now, I can see two possible solutions:

1. Run the migration process in the Server.cpp dile within the Init method after initializing of Fabric table, but before initializing of Persistent subscriptions - *this is not possible, because we need to use the* `CommonCaseDeviceServerInitParams`.
2. Run the migration process after Server initialization using Local ServerInitParams, but then restoring the subscription during Server Init will fail and we need to somehow recall restoring the subscription after the migration.

Anyway, for now, we would like to extend the Operational Credentials API to allow the migration. I've created appropriate tests for the new API, and when we will be moving to PSA API in Nrfconnect examples we will propose a new solution.

@bzbarsky-apple @tcarmelveilleux What do you think about that?

bzbarsky-apple commented on src/crypto/OperationalKeystore.h:
```diff
@@ -119,6 +119,41 @@ class OperationalKeystore
+     * @param outKeypair - a reference to P256SerializedKeypair object to store the exported key.
+     * @retval CHIP_ERROR on success.
+     * @retval CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE if the key cannot be exported due to security restrictions.
+     * @retval CHIP_ERROR_NOT_IMPLEMENTED if the exporting is not implemented for the crypto engine.
```
```suggestion
     * @retval CHIP_ERROR_NOT_IMPLEMENTED if the exporting is not implemented for the cryptography backend.
```

bzbarsky-apple commented on src/crypto/OperationalKeystore.h:
```diff
@@ -119,6 +119,41 @@ class OperationalKeystore
+    virtual CHIP_ERROR ExportOpKeypairForFabric(FabricIndex fabricIndex, Crypto::P256SerializedKeypair & outKeypair) = 0;
+
+    /**
+     * @brief Migrate the operational keypair from the extern Operational keystore (`operationalKeystore`) to this one.
```
What does "extern" mean here?

bzbarsky-apple commented on src/crypto/OperationalKeystore.h:
```diff
@@ -119,6 +119,41 @@ class OperationalKeystore
+     * @retval CHIP_ERROR_BUFFER_TOO_SMALL if `outKeyPair` buffer is too small to store the read out keypair.
+     * @retval other CHIP_ERROR value on internal storage or crypto engine errors.
+     */
+    virtual CHIP_ERROR ExportOpKeypairForFabric(FabricIndex fabricIndex, Crypto::P256SerializedKeypair & outKeypair) = 0;
```
Shouldn't this default to a CHIP_ERROR_NOT_IMPLEMENTED implementation?  Otherwise this will break compiles for anyone updating the SDK...

bzbarsky-apple commented on src/crypto/OperationalKeystore.h:
```diff
@@ -119,6 +119,41 @@ class OperationalKeystore
+     * for Fabric
+     * @retval CHIP_ERROR on success
+     * @retval CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE if the key cannot be exported due to security restrictions.
+     * @retval CHIP_ERROR_NOT_IMPLEMENTED if the exporting is not implemented for the crypto engine.
```
```suggestion
     * @retval CHIP_ERROR_NOT_IMPLEMENTED if the exporting is not implemented for the cryptography backend.
```

bzbarsky-apple commented on src/crypto/OperationalKeystore.h:
```diff
@@ -119,6 +119,41 @@ class OperationalKeystore
+     * @retval CHIP_ERROR_BUFFER_TOO_SMALL if `keyPair` buffer is too small to store the read out keypair.
+     * @retval other CHIP_ERROR value on internal storage or crypto engine errors.
+     */
+    virtual CHIP_ERROR MigrateOpKeypairForFabric(FabricIndex fabricIndex, OperationalKeystore & operationalKeystore) const = 0;
```
Again, should this not default to CHIP_ERROR_NOT_IMPLEMENTED?

bzbarsky-apple commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -154,6 +184,40 @@ CHIP_ERROR PSAOperationalKeystore::CommitOpKeypairForFabric(FabricIndex fabricIn
+CHIP_ERROR PSAOperationalKeystore::ExportOpKeypairForFabric(FabricIndex fabricIndex, Crypto::P256SerializedKeypair & outKeypair)
+{
+    // Currently exporting the key is forbidden in PSAOperationalKeystore because the PSA_KEY_USAGE_EXPORT usage flag is not set, so
+    // there is no need to compile the code for the device, but there should be an implementation for the test purposes to verify if
```
```suggestion
    // there is no need to compile the code for the device, but there should be an implementation for test purposes to verify if
```

tcarmelveilleux commented on src/crypto/OperationalKeystore.h:
```diff
@@ -119,6 +119,47 @@ class OperationalKeystore
+     * @retval CHIP_ERROR_NOT_IMPLEMENTED if the exporting is not implemented for the cryptography backend.
+     * @retval CHIP_ERROR_INVALID_FABRIC_INDEX if there is no keypair found for `fabricIndex`.
+     * @retval CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND if there is no keypair found for `fabricIndex`.
+     * @retval CHIP_ERROR_BUFFER_TOO_SMALL if `keyPair` buffer is too small to store the read out keypair.
```
There is not keyPair buffer

tcarmelveilleux commented on src/crypto/PSAOperationalKeystore.cpp:
```diff
@@ -209,5 +273,31 @@ void PSAOperationalKeystore::ReleasePendingKeypair()
+        // Migrated key is not useful anymore, remove it from the previous keystore.
+        ReturnErrorOnFailure(operationalKeystore.RemoveOpKeypairForFabric(fabricIndex));
+    }
+    else if (operationalKeystore.HasOpKeypairForFabric(fabricIndex))
+    {
+        ReturnErrorOnFailure(operationalKeystore.RemoveOpKeypairForFabric(fabricIndex));
+    }
```
What does the else cover? Is it tested? This just looks like it deletes a key, but only the other block above did a migration.

ArekBalysNordic replied:
If the new Operational Keystore already contains the key for a given fabric, the attempt to replace it should be forbidden, and we should remove the old one. It prevents from attack that someone can write directly the key to the obsolete keystore and in this way replace the converted one. And regarding tests, yes, it is tested in https://github.com/project-chip/connectedhomeip/pull/31279/files#diff-2e13b279258becb5810bff25892cd615de76dac46ba1ba1fbb78fcbd4450edebR369 and https://github.com/project-chip/connectedhomeip/pull/31279/files#diff-2cab20d625cc0e0133ed42f0de6b8a30e40dfd529490f80d7730e4f8cff42839R252 .

tcarmelveilleux commented on src/crypto/PersistentStorageOperationalKeystore.cpp:
```diff
@@ -86,6 +86,54 @@ CHIP_ERROR StoreOperationalKey(FabricIndex fabricIndex, PersistentStorageDelegat
+        VerifyOrReturnError(keyData.size() <= serializedOpKey.Capacity(), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+        // Before doing anything with the key, validate format further.
+        ReturnErrorOnFailure(reader.ExitContainer(containerType));
+        ReturnErrorOnFailure(reader.VerifyEndOfContainer());
```
Why verify end of container? Could this will eventually start failing if we have a new element later and the end of container is not hit?

ArekBalysNordic replied:
In this case, we don't expect a new element later, but actually, that's right we should not fail if the end container is not hit, but we should only check if the data from the container is valid.

tcarmelveilleux commented on src/crypto/PersistentStorageOperationalKeystore.cpp:
```diff
@@ -310,4 +326,36 @@ void PersistentStorageOperationalKeystore::ReleaseEphemeralKeypair(Crypto::P256K
+    }
+    else if (operationalKeystore.HasOpKeypairForFabric(fabricIndex))
+    {
+        ReturnErrorOnFailure(operationalKeystore.RemoveOpKeypairForFabric(fabricIndex));
```
Same comment as for the PSA version.
